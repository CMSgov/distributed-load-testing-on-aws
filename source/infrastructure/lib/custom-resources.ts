// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Aws, Construct, CustomResource } from '@aws-cdk/core';


/**
 * CustomResourcesConstruct props
 * @interface CustomResourcesConstructProps
 */
export interface CustomResourcesConstructProps {
    readonly apiEndpoint: string;
    //Custom Resource lambda
    readonly customResourceLambda: string;
    // Cognito Resources
    readonly cognitoIdentityPool: string;
    readonly cognitoUserPool: string;
    readonly cognitoUserPoolClient: string;
    // UI Console S3 bucket
    readonly consoleBucketName: string;
    // Test scenarios storage
    readonly scenariosBucket: string;
    // Source code details
    readonly sourceCodeBucketName: string;
    readonly sourceCodePrefix: string;
}

/**
 * @class
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function, a solution UUID, and a custom resource to send anonymous usage.
 */
export class CustomResourcesConstruct extends Construct {

    constructor(scope: Construct, id: string, props: CustomResourcesConstructProps) {
        super(scope, id);

        new CustomResource(this, 'CopyConsoleFiles', {
            serviceToken: props.customResourceLambda,
            resourceType: 'Custom::CopyConsoleFiles',
            properties: {
                Resource: 'CopyAssets',
                SrcBucket: props.sourceCodeBucketName,
                SrcPath: `${props.sourceCodePrefix}/console`,
                ManifestFile: 'console-manifest.json',
                DestBucket: props.consoleBucketName
            }
        });

        const awsExports = `const awsConfig = {
            cw_dashboard: 'https://console.aws.amazon.com/cloudwatch/home?region=${Aws.REGION}#dashboards:name=',
            ecs_dashboard: 'https://${Aws.REGION}.console.aws.amazon.com/ecs/home?region=${Aws.REGION}#/clusters/${Aws.STACK_NAME}/tasks',
            aws_project_region: '${Aws.REGION}',
            aws_cognito_region: '${Aws.REGION}',
            aws_cognito_identity_pool_id: '${props.cognitoIdentityPool}',
            aws_user_pools_id: '${props.cognitoUserPool}',
            aws_user_pools_web_client_id: '${props.cognitoUserPoolClient}',
            oauth: {},
            aws_cloud_logic_custom: [
                {
                    name: 'dlts',
                    endpoint: '${props.apiEndpoint}',
                    region: '${Aws.REGION}'
                }
            ],
            aws_user_files_s3_bucket: '${props.scenariosBucket}',
            aws_user_files_s3_bucket_region: '${Aws.REGION}'
        }`;

        new CustomResource(this, 'ConsoleConfig', {
            serviceToken: props.customResourceLambda,
            resourceType: 'Custom::CopyConfigFiles',
            properties: {
                Resource: 'ConfigFile',
                DestBucket: props.consoleBucketName,
                AwsExports: awsExports
            }
        });
    }
}