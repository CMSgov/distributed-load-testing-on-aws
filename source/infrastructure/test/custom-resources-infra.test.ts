// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CustomResourceInfraConstruct } from '../lib/custom-resources/custom-resources-infra';


test('DLT API Test', () => {

  const stack = new Stack();
  const testSourceBucket = new Bucket(stack, 'testSourceCodeBucket');
  const testPolicy = new Policy(stack, 'TestPolicy', {
    statements: [
      new PolicyStatement({
        resources: ['*'],
        actions: ['cloudwatch:Get*']
      })
    ]
  });

  new CustomResourceInfraConstruct(stack, 'TestCustomResourceInfra', {
    cloudWatchPolicy: testPolicy,
    consoleBucketArn: 'test:console:bucket:arn',
    mainStackRegion: 'test-region-1',
    metricsUrl: 'http://testurl.com',
    scenariosS3Bucket: 'scenariotestbucket',
    scenariosTable: 'scenarioTestTable',
    solutionId: 'S0XXX',
    solutionVersion: 'testVersion',
    sourceCodeBucket: testSourceBucket,
    sourceCodePrefix: 'test/source/prefix',
    stackType: 'main'
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
    Code: {
      S3Bucket: {
        "Ref": "testSourceCodeBucketC577B176",
      },
      S3Key: "test/source/prefix/main-custom-resource.zip",
    },
    Description: "CFN Lambda backed custom resource to deploy assets to s3",
    Environment: {
      Variables: {
        DDB_TABLE: "scenarioTestTable",
        MAIN_REGION: "test-region-1",
        METRIC_URL: "http://testurl.com",
        S3_BUCKET: "scenariotestbucket",
        SOLUTION_ID: "S0XXX",
        VERSION: "testVersion",
      }
    },
    Handler: "index.handler",
    Role: {
      "Fn::GetAtt": [
        "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
        "Arn",
      ],
    },
    Runtime: "nodejs14.x",
    Timeout: 120
  });
});
