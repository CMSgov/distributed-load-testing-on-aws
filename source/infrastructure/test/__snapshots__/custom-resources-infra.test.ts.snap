// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
Object {
  "Resources": Object {
    "TestCustomResourceInfraCustomResourceLambda92E822A7": Object {
      "DependsOn": Array [
        "TestCustomResourceInfraCustomResourceLambdaRoleDefaultPolicy46B2072F",
        "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W58",
              "reason": "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
            },
            Object {
              "id": "W89",
              "reason": "VPC not needed for lambda",
            },
            Object {
              "id": "W92",
              "reason": "Does not run concurrent executions",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "testSourceCodeBucketC577B176",
          },
          "S3Key": "test/source/prefix/main-custom-resource.zip",
        },
        "Description": "CFN Lambda backed custom resource to deploy assets to s3",
        "Environment": Object {
          "Variables": Object {
            "DDB_TABLE": "scenarioTestTable",
            "MAIN_REGION": "test-region-1",
            "METRIC_URL": "http://testurl.com",
            "S3_BUCKET": "scenariotestbucket",
            "SOLUTION_ID": "S0XXX",
            "VERSION": "testVersion",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestCustomResourceInfraCustomResourceLambdaRole03671AE8": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "iot:DescribeEndpoint and iot:DetachPrincipalPolicy cannot specify the resource.",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "testSourceCodeBucketC577B176",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                },
                Object {
                  "Action": Array [
                    "s3:PutObject",
                    "s3:DeleteObject",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":s3:::scenariotestbucket/*",
                      ],
                    ],
                  },
                },
                Object {
                  "Action": Array [
                    "dynamodb:PutItem",
                    "dynamodb:DeleteItem",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":dynamodb:test-region-1:",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":table/scenarioTestTable",
                      ],
                    ],
                  },
                },
                Object {
                  "Action": Array [
                    "iot:DescribeEndpoint",
                    "iot:DetachPrincipalPolicy",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
                Object {
                  "Action": "iot:ListTargetsForPolicy",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iot:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":policy/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CustomResourcePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestCustomResourceInfraCustomResourceLambdaRoleDefaultPolicy46B2072F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": Array [
                "test:console:bucket:arn",
                "test:console:bucket:arn/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestCustomResourceInfraCustomResourceLambdaRoleDefaultPolicy46B2072F",
        "Roles": Array [
          Object {
            "Ref": "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestPolicyCC05E598": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "cloudwatch:Get*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestPolicyCC05E598",
        "Roles": Array [
          Object {
            "Ref": "TestCustomResourceInfraCustomResourceLambdaRole03671AE8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testSourceCodeBucketC577B176": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
